src/test/java/reactor/ipc/netty/http/client/HttpClientWSOperationsTest.java

```diff
@@ -0,0 +1,119 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.ipc.netty.http.client;
+
+import java.time.Duration;
+
+import org.hamcrest.CoreMatchers;
+import org.junit.Assert;
+import org.junit.Test;
+
+import reactor.core.publisher.Mono;
+import reactor.ipc.netty.NettyContext;
+import reactor.ipc.netty.http.server.HttpServer;
+import reactor.ipc.netty.http.websocket.WebsocketOutbound;
+import reactor.test.StepVerifier;
+
+/**
+ * @author Violeta Georgieva
+ */
+public class HttpClientWSOperationsTest {
+
+	@Test
+	public void failOnClientErrorDisabled() {
+		failOnClientServerError(false, false, 401, "", "");
+	}
+
+	@Test
+	public void failOnClientErrorEnabled() {
+		failOnClientServerError(true, false, 401, "", "");
+	}
+
+	@Test
+	public void failOnServerErrorDisabled() {
+		failOnClientServerError(false, false, 500, "", "");
+	}
+
+	@Test
+	public void failOnServerErrorEnabled() {
+		failOnClientServerError(false, true, 500, "", "");
+	}
+
+	@Test
+	public void failOnClientServerErrorDisabledFailedNegotiation() {
+		failOnClientServerError(false, false, 200, "Server-Protocol", "Client-Protocol");
+	}
+
+	private void failOnClientServerError(boolean clientError, boolean serverError,
+			int serverStatus, String serverSubprotocol, String clientSubprotocol) {
+		NettyContext httpServer = HttpServer.create(0).newRouter(
+			routes -> routes.post("/login", (req, res) -> {
+						res.status(serverStatus);
+						return res.sendHeaders();
+					})
+					.get("/ws", (req, res) -> {
+						int token = Integer.parseInt(req.requestHeaders().get("Authorization"));
+						if (token >= 400) {
+							res.status(token);
+							return res.send();
+						}
+						return res.sendWebsocket(serverSubprotocol, (i, o) -> o.sendString(Mono.just("test")));
+					})
+			)
+			.block(Duration.ofSeconds(30));
+
+		Mono<HttpClientResponse> response =
+			HttpClient.create(httpServer.address().getPort())
+			          .get("/ws", request -> Mono.just(request.failOnClientError(clientError)
+			                                                  .failOnServerError(serverError))
+			                                     .transform(req -> doLoginFirst(req, httpServer.address().getPort()))
+			                                     .flatMapMany(req -> ws(req, clientSubprotocol)))
+			          .switchIfEmpty(Mono.error(new Exception()));
+
+
+		if (clientError || serverError || (!serverSubprotocol.isEmpty() && !clientSubprotocol.isEmpty())) {
+			StepVerifier.create(response)
+			            .expectError()
+			            .verify(Duration.ofSeconds(30));
+		}
+		else {
+			StepVerifier.create(response)
+			            .assertNext(res -> Assert.assertThat(res.status().code(), CoreMatchers.is(serverStatus)))
```

Comment:
(minor) we now favor AssertJ library for assertions like this. (static import `Assertions.assertThat` and replace line with `.assertNext(res -> assertThat(res.status().code()).isEqualTo(serverStatus))`)